util/signature/verifier.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package signature
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/offchainlabs/nitro/util/contracts"
+	"github.com/pkg/errors"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type Verifier struct {
+	requireSignature bool
+	authorizedMap    map[common.Address]struct{}
+	bpValidator      contracts.BatchPosterVerifierInterface
+}
+
+func NewVerifier(requireSignature bool, authorizedAddresses []common.Address, bpValidator contracts.BatchPosterVerifierInterface) *Verifier {
+	authorizedMap := make(map[common.Address]struct{}, len(authorizedAddresses))
+	for _, addr := range authorizedAddresses {
+		authorizedMap[addr] = struct{}{}
+	}
+	return &Verifier{
+		requireSignature: requireSignature,
+		authorizedMap:    authorizedMap,
+		bpValidator:      bpValidator,
+	}
+}
+
+func (v *Verifier) VerifyHash(ctx context.Context, signature []byte, hash common.Hash) (bool, error) {
+	return v.VerifyClosure(ctx, signature, func() common.Hash { return hash })
+}
+
+func (v *Verifier) VerifyData(ctx context.Context, signature []byte, data ...[]byte) (bool, error) {
+	return v.VerifyClosure(ctx, signature, func() common.Hash { return crypto.Keccak256Hash(data...) })
+}
+
+func (v *Verifier) VerifyClosure(ctx context.Context, signature []byte, getHash func() common.Hash) (bool, error) {
+	if signature == nil {
```

Comment:
We should be careful here as `[]byte{} != nil`. Checking `if len(signature) == 0` might be safer.